checker/cost.go

```diff
@@ -0,0 +1,430 @@
+	case *exprpb.Expr_CallExpr:
+		cost = c.costCall(e)
+	case *exprpb.Expr_ListExpr:
+		cost = c.costCreateList(e)
+	case *exprpb.Expr_StructExpr:
+		cost = c.costCreateStruct(e)
+	case *exprpb.Expr_ComprehensionExpr:
+		cost = c.costComprehension(e)
+	default:
+		return CostEstimate{}
+	}
+	return cost
+}
+
+func (c *coster) costIdent(e *exprpb.Expr) CostEstimate {
+	identExpr := e.GetIdentExpr()
+
+	// resolve the field path
+	if ident := c.env.LookupIdent(identExpr.GetName()); ident != nil {
+		if binding, ok := c.iterVarMapping[ident]; ok {
+			c.exprPath[e.Id] = append(c.exprPath[binding], "*")
+		} else {
+			c.exprPath[e.Id] = []string{identExpr.GetName()}
+		}
+
+		return identCost
+	}
+	return CostEstimate{}
+}
+
+func (c *coster) costSelect(e *exprpb.Expr) CostEstimate {
+	sel := e.GetSelectExpr()
+	var sum CostEstimate
+
+	if sel.TestOnly {
+		return sum
+	}
+	sum = sum.Add(c.cost(sel.Operand))
+	c.exprPath[e.Id] = append(c.exprPath[sel.Operand.Id], sel.Field)
+	targetType := c.getType(sel.Operand)
+	switch kindOf(targetType) {
+	case kindMap, kindObject, kindTypeParam:
+		sum = sum.Add(selectCost)
+	}
+	return sum
+}
+
+func (c *coster) costCall(e *exprpb.Expr) CostEstimate {
+	call := e.GetCallExpr()
+	target := call.GetTarget()
+	args := call.GetArgs()
+
+	var sum CostEstimate
+
+	argTypes := make([]AstNode, len(args))
+	for i, arg := range args {
+		// TODO: && || short circuit, so min cost should only include 1st arg eval
+		// unless exhaustive evaluation is enabled
+		sum = sum.Add(c.cost(arg))
+		argTypes[i] = c.newAstNode(arg)
+	}
+
+	ref := c.checker.ReferenceMap[e.Id]
+	if ref == nil || len(ref.GetOverloadId()) == 0 {
+		return CostEstimate{}
+	}
+	var targetType AstNode
+	if target != nil {
+		if call.Target != nil {
+			sum = sum.Add(c.cost(call.Target))
+			targetType = c.newAstNode(call.Target)
+		}
+	}
+	// Pick a cost estimate range that covers all the overload cost estimation ranges
+	fnCost := CostEstimate{Min: uint64(math.MaxUint64), Max: 0}
+	for _, overload := range ref.GetOverloadId() {
+		overloadCost := c.functionCost(overload, &targetType, argTypes)
+		if overloadCost.Max > fnCost.Max {
+			fnCost.Max = overloadCost.Max
+		}
+		if overloadCost.Min < fnCost.Min {
+			fnCost.Min = overloadCost.Min
+		}
+	}
+	return sum.Add(fnCost)
+}
+
+func (c *coster) costCreateList(e *exprpb.Expr) CostEstimate {
+	create := e.GetListExpr()
+	var sum CostEstimate
+	for _, e := range create.Elements {
+		sum = sum.Add(c.cost(e))
+	}
+	return sum.Add(createListBaseCost)
+}
+
+func (c *coster) costCreateStruct(e *exprpb.Expr) CostEstimate {
+	str := e.GetStructExpr()
+	if str.MessageName != "" {
+		return c.costCreateMessage(e)
+	} else {
+		return c.costCreateMap(e)
+	}
+}
+
+func (c *coster) costCreateMap(e *exprpb.Expr) CostEstimate {
+	mapVal := e.GetStructExpr()
+	var sum CostEstimate
+	for _, ent := range mapVal.GetEntries() {
+		key := ent.GetMapKey()
+		sum = sum.Add(c.cost(key))
+
+		sum = sum.Add(c.cost(ent.Value))
```

Comment:
Try to consistently use the getter methods here and elsewhere on proto field accesses. I believe at some point only getters will be supported, e.g. `ent.GetValue()`

Reply:
I'll given it the college try :)